compiler/pipes/calc-bad-vars.cpp

```diff
@@ -115,13 +115,297 @@ struct FuncCallGraph {
   }
 };
 
+class CallstackOfColoredFunctions {
+  std::vector<FunctionPtr> stack;                       // functions placed in order, only colored functions
+  std::vector<function_palette::color_t> colors_chain;  // blended colors of stacked functions, one-by-one
+  std::unordered_set<FunctionPtr> index_set;            // a quick index for contains(), has the same elements as stack
+  function_palette::colors_mask_t colors_mask{0};       // mask of all colors_chain
+
+  void recalc_mask() {
+    colors_mask = 0;
+    for (function_palette::color_t c : colors_chain) {
+      colors_mask |= c;
+    }
+  }
+
+public:
+  CallstackOfColoredFunctions() {
+    stack.reserve(10);
+    index_set.reserve(10);
+  }
+
+  size_t size() const {
+    return stack.size();
+  }
+
+  const std::vector<FunctionPtr> &as_vector() const {
+    return stack;
+  }
+
+  const std::vector<function_palette::color_t> &get_colors_chain() const {
+    return colors_chain;
+  }
+
+  function_palette::colors_mask_t get_colors_mask() const {
+    return colors_mask;
+  }
+
+  void push_back(const FunctionPtr &el) {
+    stack.push_back(el);
+    index_set.insert(el);
+    for (const auto &sep_color : el->colors) {
+      colors_chain.push_back(sep_color);
+    }
+    recalc_mask();
+  }
+
+  void pop_back() {
+    FunctionPtr pop = stack.back();
+    index_set.erase(pop);
+    stack.pop_back();
+    for (const auto &sep_color __attribute__((unused)) : pop->colors) {
+      colors_chain.pop_back();
```

Comment:
`colors_chain.resize(colors_chain.size() - pop->colors.size());`?

Reply:
Hm, I didn't know, that it works ) I'll test. If ok — I'll replace with your suggestion

Reply:
pop->colors is a forward_list, so calculating its size will result in the same loop